ocw/lib/gce.py

```diff
@@ -207,3 +209,112 @@ def cleanup_all(self):
                 if "warnings" in response:
                     for w in response["warnings"]:
                         self.log_warn(w["message"])
+        self.cleanup_vault_keys()
+
+    def cleanup_vault_keys(self):
+        self.log_info("## GCE Vault key cleaning - %s ##" % self.__project)
+        credentials = service_account.Credentials.from_service_account_info(
+            self.__credentials.getPrivateKeyData()
+        )
+        service = googleapiclient.discovery.build('iam', 'v1', credentials=credentials)
+        # now get service accounts
+        vault_emails = self.parser_service_account_list(service)
+        # for each vault account find the associated names with old keys
+        vault_keys = self.get_vault_name(service, vault_emails)
+
+        self.delete_vault_keys(service, vault_keys)
+
+    @filterAfterTime(hours=24)
```

Comment:
can you elaborate benefit of using decorator pattern here ? AFAIK decorator should be used when you need to do same thing *in several classes not linked by inheritance* than you can save yourself from *code duplication* . Can not find anything related to this topic here . In a fact you just break single flow into two pieces and now it is hard to understand what actually going on with data ...

Reply:
to be honest, i dont know how to answer this question and i try to avoid them in general (many reason out of the scope here). Everyone should do their research. for instance decorators do not apply _in several classes_ only. But i will tell you that. some small portion of my thinking is that `filterAfterTime` will apply to ec2 and azure later. Unfortunately i dont have idea how it will look like there so i do not provide the one-off solution. i expect that it will require some changes when i will have clear picture from the other providers.

> In a fact you just break single flow into two pieces and now it is hard to understand what actually going on with data ...

Thats not true. in fact this design is very common in many frameworks (django, Spring, etc). the name of the functions is also self-explanatory. The `get_vault_name` requests and returns the `['keys']` and then you want to `filterAfterTime`, to get the _old_ ones.  `get_vault_name` will never have to change. if believe that if you see what `filterAfterTime` does you will notice that import different packages and does quite different operation, which do not belong inside `get_vault_name` as you would expect from such function.

Reply:
> to be honest, i dont know how to answer this question and i try to avoid them in general (many reason out of the scope here)

sorry I didn't get this part 

> some small portion of my thinking is that filterAfterTime will apply to ec2 and azure later.

currently decorators are not used in this project at all . this PR is about adding google service account cleanup . I would prefer that we will concentrate on this without doing several things at the same time . I am fine to add decorators to the project but I can not accept having this pattern in one single random place while whole project not using same approach . Please solve : 
> Unfortunately i dont have idea how it will look like 

and provide " one-off solution" in separate PR.

Reply:
*changes applied and waiting for feedback*